src/persistent_tasks.jl

```diff
@@ -0,0 +1,156 @@
+"""
+    Aqua.test_persistent_tasks(package; tmax=5)
+
+Test whether loading `package` creates persistent `Task`s
+which may block precompilation of dependent packages.
+
+# Motivation
+
+Julia 1.10 and higher wait for all running `Task`s to finish
+before writing out the precompiled (cached) version of the package.
+One consequence is that a package that launches
+`Task`s in its `__init__` function may precompile successfully,
+but block precompilation of any packages that depend on it.
+
+# Example
+
+Let's create a dummy package, `PkgA`, that launches a persistent `Task`:
+
+```julia
+module PkgA
+const t = Ref{Any}()   # to prevent the Timer from being garbage-collected
+__init__() = t[] = Timer(0.1; interval=1)   # create a persistent `Timer` `Task`
+end
+```
+
+`PkgA` will precompile successfully, because `PkgA.__init__()` does not
+run when `PkgA` is precompiled. However,
+
+```julia
+module PkgB
+using PkgA
+end
+```
+
+fails to precompile: `using PkgA` runs `PkgA.__init__()`, which
+leaves the `Timer` `Task` running, and that causes precompilation
+of `PkgB` to hang.
+
+# How the test works
+
+This test works by launching a Julia process that tries to precompile a
+dummy package similar to `PkgB` above, modified to signal back to Aqua when
+`PkgA` has finished loading. The test fails if the gap between loading `PkgA`
+and finishing precompilation exceeds time `tmax`.
+
+# How to fix failing packages
+
+Often, the easiest fix is to modify the `__init__` function to check whether the
+Julia process is precompiling some other package; if so, don't launch the
+persistent `Task`s.
+
+```
+function __init__()
+    # Other setup code here
+    if ccall(:jl_generating_output, Cint, ()) == 0   # if we're not precompiling...
+        # launch persistent tasks here
+    end
+end
+```
+
+In more complex cases, you may need to set up independently-callable functions
+to launch the tasks and cleanly shut them down.
+
+# Arguments
+- `package`: a top-level `Module` or `Base.PkgId`.
+
+# Keyword Arguments
+- `tmax::Real`: the maximum time (in seconds) to wait after loading the
+  package before forcibly shutting down the precompilation process (triggering
+  a test failure).
+"""
+function test_persistent_tasks(package::PkgId; tmax = 10, fails::Bool = false)
+    @testset "$package persistent_tasks" begin
+        result = root_project_or_failed_lazytest(package)
+        result isa LazyTestResult && return result
+        @test fails ⊻ precompile_wrapper(result, tmax)
```

Comment:
So perhaps like this (but obviously callers would still need to be adjusted)
```suggestion
- `broken::Bool = false`: If true, it uses `@test_broken` instead of
  `@test`.
"""
function test_persistent_tasks(package::PkgId; tmax = 10, broken::Bool = false)
    @testset "$package persistent_tasks" begin
        result = root_project_or_failed_lazytest(package)
        result isa LazyTestResult && return result
        if broken
            @test_broken precompile_wrapper(result, tmax)
        else
            @test precompile_wrapper(result, tmax)
        end
```

Reply:
I merged this, but then I changed it back to the `⊻`. The issue is that `@test_broken` usually marks something you want to fix in the package being tested, or a bug in Julia itself. In this case, it's more like `@test_throws` and we actually want it to fail---that's a sign that Aqua is doing its job correctly. So the xor logic seems correct.

Reply:
Hmm, now I worry: was I right to use something different from `broken` at the outset? Is `broken` really intended for users of Aqua? `fails` was only there for testing, it's not intended to be used to mark tests in other packages as things you'd like to have pass, but don't.